### PR TITLE
Add a connection timeout

### PR DESCRIPTION
**Pull request by Jmgr**
This adds a connection timeout used by Realtime, REST, and SSE connections.

### CODE REVIEW

**Review by lmars — COMMENTED** [review]
I think we should make a distinction between a connection timeout (i.e. how long it takes to establish a connection), and a request timeout (i.e. how long it takes for a HTTP request to fully complete), since I would expect connections to be established in a smaller time frame than requests completing.

The Ably SDK has this distinction (see [here](https://docs.ably.io/client-lib-development-guide/features/#TO3l)), so let's just map through to those in ably-go.

**Comment by Jmgr** [comment]
I agree. Looking at `ably-go` with tag `v1.2.0-apipreview.4` I can indeed see a few options:
- `RealtimeRequestTimeout` that is used for the initial Dial and receiving events
- `HTTPOpenTimeout` that is not used anywhere
- `HTTPRequestTimeout` that is passed down to http.Client but cannot be set (missing `WithX` function)
So I think this needs some work on `ably-go` first. I'll create a ticket if there is none yet.
In the meantime using `WithHTTPClient` as a workaround should work, and we can update `ably-boomer` once a new version of `ably-go` comes out. WDYT?

**Comment by lmars** [comment]
> So I think this needs some work on ably-go first. I'll create a ticket if there is none yet.

Sure, although I think it would be _really_ helpful to make a small PR instead 😉 

> In the meantime using WithHTTPClient as a workaround should work, and we can update ably-boomer once a new version of ably-go comes out. WDYT?

SGTM 👍

**Comment by Jmgr** [comment]
Coming back to this:
In `ably-go`, `RealtimeRequestTimeout` is used for Realtime connections and `HTTPRequestTimeout` for REST requests. Looking at the spec that seems expected. I don't really see in what cases `HTTPOpenTimeout` should be used, as REST requests don't maintain a connection.
So my fix would be to create a `WithHTTPRequestTimeout` PR that would then be used for REST requests.

In `ably-boomer` we are using REST requests for device registration and subscription, so there is no connection, only a request timeout. There is also a Realtime connection that is open, and this indeed has a connection timeout, but `RealtimeRequestTimeout` is used there. I can indeed expose in `ably-boomer` a HTTP timeout and a Realtime request timeout. Or do you mean that `ably-go` should somehow use a different timeout for establishing the initial Realtime connection than `RealtimeRequestTimeout`?

**Comment by Jmgr** [comment]
Also the Realtime connection API exposed by `ably-go` does not provide a context to cancel the connection unfortunately.

**Comment by lmars** [comment]
So from a high level, there should be two boomer configuration options:

1. Connect timeout - the maximum amount of time to wait for a TCP connection to be established
2. Request timeout - the maximum amount of time to wait for a request to complete and return a response

We'll likely want the former to be a smaller number than the latter.

Now, for the implementation, I'd suggest mapping those to the equivalent from the spec, which seems to imply:

1. Connect timeout - `httpOpenTimeout`
2. Request timeout - `httpRequestTimeout` (REST) and `realtimeRequestTimeout` (Realtime)

If ably-go isn't spec compliant, then we should make it so (i.e. these three should all be `WithXXX` options and actually be used), and then map boomer config options to those Ably client options.

> In ably-boomer we are using REST requests for device registration and subscription, so there is no connection, only a request timeout

To be clear, there _is_ a connection with a REST request (i.e. the underlying TCP connection the request is sent over), and we should set a connect timeout on it, but it's not as important as the request timeout if it's difficult to configure.

**Comment by Jmgr** [comment]
Then I think the [spec](https://docs.ably.io/client-lib-development-guide/features/#DF1b) could need an update, I could be misunderstanding it, but it says `realtimeRequestTimeout: When a realtime client library is establishing a connection with Ably [...]` so that looks like it should be using that timeout for the first connection, not `httpOpenTimeout`.

**Comment by Jmgr** [comment]
Why is `httpOpenTimeout` not `tcpOpenTimeout` if it should also be used for the Realtime connection?

**Comment by lmars** [comment]
@Jmgr to be honest, I don't know, it does seem a little ambiguous. Might be worth opening an issue on the docs repo to discuss that, but let's just do something reasonable that at least lets us set these two timeouts in the boomer config.

**Comment by Jmgr** [comment]
This should work now @lmars.

`ably-go`'s `main` branch has diverged somewhat though, notably with the removal of the `HTTPOpenTimeout` parameter. The API is also different, so there is a bit of work to rebase the `integration/1.2` branch to `main` there and get `ably-boomer` to compile with it. I haven't do that yet as I wanted to make minimal changes only but we should consider it at some point.

**Comment by lmars** [comment]
@Jmgr the ably-go `integration/1.2` branch is basically a rewrite of the API in preparation for v1.2 landing on `main`, so don't worry about `main` for now, let's just track the v1.2 branch until it's landed.

File: config/config.go
Comment (lmars) [inline config/config.go:43]: I think these should default to 4s and 10s respectively to match the default Ably SDK values

**Review by lmars — APPROVED** [review]
LGTM pending minor comment

**Pull request merged**
